services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/extension/fixture/rule/CleanupSwitchesRule.groovy

```diff
@@ -25,24 +27,66 @@ class CleanupSwitchesRule extends AbstractRuleExtension<CleanupSwitches> {
 
     @Override
     void afterTest() {
-        log.debug "Verifying that all the switches are clean"
-        try {
-            Wrappers.wait(WAIT_OFFSET) {
-                topology.activeSwitches.each { sw ->
+        topology.activeSwitches.each { sw ->
+            try {
+                Wrappers.wait(WAIT_OFFSET) {
                     def rules = northbound.validateSwitchRules(sw.dpId)
-                    assert rules.missingRules.empty
-                    assert rules.excessRules.empty
-                    assert rules.properRules.empty
+                    rules.excessRules.empty ?: failWith(new ExcessRulesException(
+                            "$sw.dpId switch appear to have excess rules at the end of the test.", rules))
+                    rules.properRules.empty ?: failWith(new ProperRulesException(
+                            "FATAL: there are proper rules left after the test. This usually means" +
+                                    " that certain flow was not properly removed from database and Kilda is in an" +
+                                    " inconsistent state. Aborting.\nswitch: $sw.dpId", rules))
+                    rules.missingRules.empty ?: failWith(new MissingRulesException(
+                            "FATAL: there are missing rules left after the test. This usually means" +
+                                    " that certain flow was not properly removed from database and Kilda is in an" +
+                                    " inconsistent state. Aborting.\nswitch: $sw.dpId", rules))
+                }
+            } catch (WaitTimeoutException e) {
+                //TODO(rtretiak): Not able to deal with randomly appearing rules from previous tests at this point.
+                //Waiting for new Kilda patch with Hub&Spoke features
+                def originalError = e.originalError
+                switch (originalError) {
+                    case ExcessRulesException:
+                        log.warn("Force switch cleanup.", originalError)
```

Comment:
I think it is not a good idea to print the original error here. It will be a long traceback that will be useless and we will not see the main message related to the switch that has excess rules. Probably we can do something like that:
log.warn(e.message + "\n\nForce cleanup for switch '$sw.dpId'\n\n")

Reply:
will change